src/main/java/org/jabref/logic/ai/embeddings/MVStoreEmbeddingStore.java

```diff
@@ -0,0 +1,163 @@
+package org.jabref.logic.ai.embeddings;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.UUID;
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import dev.langchain4j.data.document.Metadata;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.store.embedding.CosineSimilarity;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
+import dev.langchain4j.store.embedding.EmbeddingSearchResult;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import dev.langchain4j.store.embedding.RelevanceScore;
+import dev.langchain4j.store.embedding.filter.Filter;
+import dev.langchain4j.store.embedding.filter.comparison.IsEqualTo;
+import dev.langchain4j.store.embedding.filter.comparison.IsIn;
+import jakarta.annotation.Nullable;
+import org.h2.mvstore.MVMap;
+import org.h2.mvstore.MVStore;
+
+import static java.util.Comparator.comparingDouble;
+
+/**
+ * A custom implementation of langchain4j's {@link EmbeddingStore} that uses a {@link MVStore} as an embedded database.
+ */
+public class MVStoreEmbeddingStore implements EmbeddingStore<TextSegment> {
+    private final MVMap<String, float[]> embeddingsMap;
+    private final MVMap<String, String> fileMap;
+    private final MVMap<String, String> contentsMap;
+
+    public MVStoreEmbeddingStore(MVStore mvStore) {
+        // TODO: Will this work efficiently? Does optimizations work on map level or entry level?
+
+        this.embeddingsMap = mvStore.openMap("embeddingsMap");
+        this.fileMap = mvStore.openMap("fileMap");
+        this.contentsMap = mvStore.openMap("contentsMap");
+    }
+
+    @Override
+    public String add(Embedding embedding) {
+        String id = String.valueOf(UUID.randomUUID());
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        embeddingsMap.put(id, embedding.vector());
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = String.valueOf(UUID.randomUUID());
+        add(id, embedding);
+
+        contentsMap.put(id, textSegment.text());
+
+        String linkedFile = textSegment.metadata().getString("linkedFile");
```

Comment:
Usually all embeddings come from files, and so in current version of JabRef every embedding will have a `"linkedFile"` metadata entry.

It's not an error to have an embedding without this entry, though it won't be useful, because chats filter embeddings by files, and so that embedding without linked file will be omitted.

Let's just log it